File: crates/mlsirm-core/src/mmle.rs
//! MMLE (marginal maximum likelihood) via EM for a unidimensional 2PL model.
//!
//! Penalized JMLE estimates each person's ability jointly with item parameters,
//! which is weakly identified and biased under missing/sparse data. MMLE
//! integrates ability out over a population distribution (Gauss-Hermite
//! quadrature), so each person contributes a product over their **observed**
//! items only — missingness (MAR) is handled by construction, no imputation.
//!
//! Scope: unidimensional 2PL, logit = a*theta + b, ability prior N(0, 1).
//! Item parameters (a, b) by EM; ability returned as EAP posterior mean.

/// 41-node probabilists' Gauss-Hermite rule, weights normalized to sum 1
/// (quadrature for a standard-normal N(0,1) ability prior). Values are the
/// shortest-roundtrip f64 output of `numpy.polynomial.hermite_e.hermegauss(41)`
/// (weights divided by their sum), so this table is bit-identical to the
/// default quadrature of the NumPy reference in
/// `python/fast_mlsirm/estimators/mmle.py` — the Rust<->NumPy parity contract.
pub(crate) const GH_NODES: [f64; 41] = [
    -11.614937254337464,
    -10.647536786319334,
    -9.843433249157995,
    -9.123069907984473,
    -8.45609908326939,
    -7.82688200405387,
    -7.226022663732788,
    -6.647308470747189,
    -6.0863491648784755,
    -5.539884440458124,
    -5.0053966834041255,
    -4.480878331594007,
    -3.9646840280332665,
    -3.4554322177809933,
    -2.9519370163811907,
    -2.453159345907048,
    -1.9581707119772913,
    -1.4661254572959665,
    -0.9762387671800493,
    -0.4877685693194346,
    0.0,
    0.4877685693194346,
    0.9762387671800493,
    1.4661254572959665,
    1.9581707119772913,
    2.453159345907048,
    2.9519370163811907,
    3.4554322177809933,
    3.9646840280332665,
    4.480878331594007,
    5.0053966834041255,
    5.539884440458124,
    6.0863491648784755,
    6.647308470747189,
    7.226022663732788,
    7.82688200405387,
    8.45609908326939,
    9.123069907984473,
    9.843433249157995,
    10.647536786319334,
    11.614937254337464,
];
pub(crate) const GH_WEIGHTS: [f64; 41] = [
    2.2578639565831077e-30,
    8.308558938782659e-26,
    2.7468912285223205e-22,
    2.3263841455871947e-19,
    7.655982291966907e-17,
    1.2203348742027809e-14,
    1.0778183949358929e-12,
    5.7698534280921236e-11,
    1.994794756757345e-9,
    4.66734770810732e-8,
    7.658186077982326e-7,
    9.058608622432971e-6,
    7.89471931950462e-5,
    0.000515801444343186,
    0.002561642428649783,
    0.009777902738208262,
    0.028937211747934403,
    0.06684765935446638,
    0.12114891701151059,
    0.17284953105060138,
    0.19454502775360044,
    0.17284953105060138,
    0.12114891701151059,
    0.06684765935446638,
    0.028937211747934403,
    0.009777902738208262,
    0.002561642428649783,
    0.000515801444343186,
    7.89471931950462e-5,
    9.058608622432971e-6,
    7.658186077982326e-7,
    4.66734770810732e-8,
    1.994794756757345e-9,
    5.7698534280921236e-11,
    1.0778183949358929e-12,
    1.2203348742027809e-14,
    7.655982291966907e-17,
    2.3263841455871947e-19,
    2.7468912285223205e-22,
    8.308558938782659e-26,
    2.2578639565831077e-30,
];

#[derive(Clone, Debug)]
pub struct MmleResult {
    pub a: Vec<f64>,
    pub b: Vec<f64>,
    pub theta: Vec<f64>,
    pub loglik_trace: Vec<f64>,
    pub n_iter: usize,
    pub converged: bool,
}

#[derive(Clone, Copy, Debug)]
pub struct MmleConfig {
    pub max_iter: usize,
    pub tol: f64,
    pub ridge_a: f64,
    pub ridge_b: f64,
    pub newton_iter: usize,
}

impl Default for MmleConfig {
    fn default() -> Self {
        Self {
            max_iter: 500,
            tol: 1e-6,
            ridge_a: 1e-3,
            ridge_b: 1e-3,
            newton_iter: 25,
        }
    }
}

#[inline]
pub(crate) fn log_sigmoid(x: f64) -> f64 {
    if x >= 0.0 {
        -(-x).exp().ln_1p()
    } else {
        x - x.exp().ln_1p()
    }
}

#[inline]
pub(crate) fn sigmoid_stable(x: f64) -> f64 {
    if x >= 0.0 {
        1.0 / (1.0 + (-x).exp())
    } else {
        let ex = x.exp();
        ex / (1.0 + ex)
    }
}

/// Calibrate a unidimensional 2PL by MMLE-EM under missing data.
/// `y` and `observed` are row-major `n_persons * n_items`. Missing cells (where
/// `observed[idx] == false`) are ignored.
pub fn fit_mmle_2pl(
    y: &[f64],
    observed: &[bool],
    n_persons: usize,
    n_items: usize,
    cfg: &MmleConfig,
) -> MmleResult {
    assert_eq!(y.len(), n_persons * n_items);
    assert_eq!(observed.len(), n_persons * n_items);
    let q = GH_NODES.len();
    let log_w: Vec<f64> = GH_WEIGHTS.iter().map(|w| w.ln()).collect();

    let mut a = vec![1.0_f64; n_items];
    let mut b = vec![0.0_f64; n_items];
    for i in 0..n_items {
        let mut num = 0.0;
        let mut den = 0.0;
        for p in 0..n_persons {
            let idx = p * n_items + i;
            if observed[idx] {
                num += y[idx];
                den += 1.0;
            }
        }
        let prop = if den > 0.0 {
            (num / den).clamp(0.02, 0.98)
        } else {
            0.5
        };
        b[i] = (prop / (1.0 - prop)).ln();
    }

    let mut loglik_trace: Vec<f64> = Vec::new();
    let mut posterior = vec![0.0_f64; n_persons * q];
    let mut converged = false;

    for iteration in 0..cfg.max_iter {
        let mut log_p1 = vec![0.0_f64; q * n_items];
        let mut log_p0 = vec![0.0_f64; q * n_items];
        for (qi, &node) in GH_NODES.iter().enumerate() {
            for i in 0..n_items {
                let eta = a[i] * node + b[i];
                log_p1[qi * n_items + i] = log_sigmoid(eta);
                log_p0[qi * n_items + i] = log_sigmoid(-eta);
            }
        }

        let mut total_loglik = 0.0;
        for p in 0..n_persons {
            let mut log_joint = vec![0.0_f64; q];
            for (qi, item) in log_joint.iter_mut().enumerate() {
                let mut acc = log_w[qi];
                for i in 0..n_items {
                    let idx = p * n_items + i;
                    if observed[idx] {
                        let yy = y[idx];
                        acc +=
                            yy * log_p1[qi * n_items + i] + (1.0 - yy) * log_p0[qi * n_items + i];
                    }
                }
                *item = acc;
            }
            let max_lj = log_joint.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
            let mut denom = 0.0;
            for &lj in &log_joint {
                denom += (lj - max_lj).exp();
            }
            total_loglik += max_lj + denom.ln();
            for qi in 0..q {
                posterior[p * q + qi] = (log_joint[qi] - max_lj).exp() / denom;
            }
        }
        loglik_trace.push(total_loglik);

        let mut n_iq = vec![0.0_f64; n_items * q];
        let mut r_iq = vec![0.0_f64; n_items * q];
        for p in 0..n_persons {
            for i in 0..n_items {
                let idx = p * n_items + i;
                if observed[idx] {
                    let yy = y[idx];
                    for qi in 0..q {
                        let post = posterior[p * q + qi];
                        n_iq[i * q + qi] += post;
                        r_iq[i * q + qi] += yy * post;
                    }
                }
            }
        }

        for i in 0..n_items {
            let (mut ai, mut bi) = (a[i], b[i]);
            for _ in 0..cfg.newton_iter {
                let (mut g_a, mut g_b, mut h_aa, mut h_bb, mut h_ab) = (0.0, 0.0, 0.0, 0.0, 0.0);
                for (qi, &node) in GH_NODES.iter().enumerate() {
                    let p_correct = sigmoid_stable(ai * node + bi);
                    let n = n_iq[i * q + qi];
                    let w = n * p_correct * (1.0 - p_correct);
                    let resid = r_iq[i * q + qi] - n * p_correct;
                    g_a += resid * node;
                    g_b += resid;
                    h_aa -= w * node * node;
                    h_bb -= w;
                    h_ab -= w * node;
                }
                g_a -= cfg.ridge_a * ai;
                g_b -= cfg.ridge_b * bi;
                h_aa -= cfg.ridge_a;
                h_bb -= cfg.ridge_b;
                let det = h_aa * h_bb - h_ab * h_ab;
                if det.abs() < 1e-12 {
                    break;
                }
                let da = (h_bb * g_a - h_ab * g_b) / det;
                let db = (h_aa * g_b - h_ab * g_a) / det;
                ai = (ai - da).clamp(1e-3, 10.0);
                bi -= db;
                if da.abs() + db.abs() < 1e-8 {
                    break;
                }
            }
            a[i] = ai;
            b[i] = bi;
        }

        if iteration > 0 {
            let delta = (loglik_trace[iteration] - loglik_trace[iteration - 1]).abs();
            if delta < cfg.tol {
                converged = true;
                break;
            }
        }
    }

    let mut theta = vec![0.0_f64; n_persons];
    for p in 0..n_persons {
        let mut m = 0.0;
        for (qi, &node) in GH_NODES.iter().enumerate() {
            m += posterior[p * q + qi] * node;
        }
        theta[p] = m;
    }

    let n_iter = loglik_trace.len();
    MmleResult {
        a,
        b,
        theta,
        loglik_trace,
        n_iter,
        converged,
    }
}

#[cfg(test)]
#[path = "../../../tests/unit/mmle_tests.rs"]
mod tests;

File: python/fast_mlsirm/estimators/mmle.py
"""MMLE (marginal maximum likelihood) via EM — robust to missing data.

Why this exists
---------------
Penalized JMLE estimates every person's theta as a free parameter jointly with
item parameters. Under missing / sparse responses that joint problem is weakly
identified and item parameters (especially discrimination) become biased.

MMLE marginalizes the person ability theta out of the likelihood by integrating
over a fixed population distribution (Gauss-Hermite quadrature). Each person's
contribution is a product over their **observed** items only, so missingness is
handled *by construction* — no imputation, no bias from unanswered items
(missing-at-random). This is the standard, statistically-sound way to calibrate
item parameters when data are incomplete.

Scope
-----
Unidimensional 2PL (matches ULS2PLM / the 2PL slice of MLS2PLM's measurement
part). Item parameters (a, b) are estimated by EM; person ability is returned as
the EAP posterior mean. Multidimensional / spatial (xi, zeta, tau) and polytomous
GRM are separate follow-up milestones.
"""

from __future__ import annotations

import numpy as np
from numpy.polynomial.hermite_e import hermegauss


MAX_GAUSS_HERMITE_NODES = 100
MAX_MMLE_FALLBACK_WORKSPACE_BYTES = 512 * 1024 * 1024


def _sigmoid(x: np.ndarray) -> np.ndarray:
    """Logistic sigmoid with the exponent clipped to ``[-35, 35]`` for stability."""
    return 1.0 / (1.0 + np.exp(-np.clip(x, -35.0, 35.0)))


def _validate_quadrature_node_count(n_nodes: int) -> int:
    """Return a node count within NumPy's documented tested quadrature range."""
    if isinstance(n_nodes, (bool, np.bool_)) or not isinstance(
        n_nodes, (int, np.integer)
    ):
        raise ValueError(
            f"n_nodes must be an integer in [1, {MAX_GAUSS_HERMITE_NODES}]"
        )
    validated = int(n_nodes)
    if validated < 1 or validated > MAX_GAUSS_HERMITE_NODES:
        raise ValueError(f"n_nodes must be in [1, {MAX_GAUSS_HERMITE_NODES}]")
    return validated


def _estimate_mmle_workspace_bytes(
    n_persons: int,
    n_items: int,
    n_nodes: int,
) -> int:
    """Conservatively estimate owned NumPy fallback workspace in bytes.

    The estimate covers response-sized conversion and masking arrays,
    person-by-node posterior arrays, item-by-node Newton arrays, and small
    one-dimensional state. It deliberately overestimates repository-owned
    arrays but does not claim to include caller-owned inputs or hidden BLAS
    workspace.
    """
    response_cells = n_persons * n_items
    person_node_cells = n_persons * n_nodes
    item_node_cells = n_items * n_nodes
    float64_cells = (
        5 * response_cells
        + 6 * person_node_cells
        + 14 * item_node_cells
        + 12 * (n_persons + n_items + n_nodes)
    )
    boolean_cells = response_cells + 2 * n_items
    return (
        float64_cells * np.dtype(np.float64).itemsize
        + boolean_cells * np.dtype(np.bool_).itemsize
    )


def _validate_mmle_workspace(
    n_persons: int,
    n_items: int,
    n_nodes: int,
) -> None:
    """Reject fallback problems whose estimated workspace exceeds the safe cap."""
    estimated_bytes = _estimate_mmle_workspace_bytes(
        n_persons,
        n_items,
        n_nodes,
    )
    if estimated_bytes > MAX_MMLE_FALLBACK_WORKSPACE_BYTES:
        raise ValueError(
            "NumPy MMLE fallback workspace estimate "
            f"{estimated_bytes} bytes exceeds the "
            f"{MAX_MMLE_FALLBACK_WORKSPACE_BYTES}-byte safe limit; "
            "use the Rust backend or reduce the response matrix or quadrature nodes"
        )


def gauss_hermite_nodes(n_nodes: int) -> tuple[np.ndarray, np.ndarray]:
    """Nodes/weights for a standard-normal ability prior N(0, 1).

    ``hermegauss`` gives the probabilists' Hermite rule (weight exp(-x^2/2));
    normalizing the weights to sum to 1 turns them into N(0,1) quadrature.
    """
    validated_nodes = _validate_quadrature_node_count(n_nodes)
    nodes, raw_weights = hermegauss(validated_nodes)
    weights = raw_weights / raw_weights.sum()
    return nodes, weights


def fit_mmle_2pl(
    y: np.ndarray,
    observed: np.ndarray,
    *,
    n_nodes: int = 41,
    max_iter: int = 500,
    tol: float = 1e-6,
    ridge_a: float = 1e-3,
    ridge_b: float = 1e-3,
    seed: int = 1,
) -> dict[str, object]:
    """Calibrate a unidimensional 2PL by MMLE-EM under missing data.

    Parameters
    ----------
    y : (n_persons, n_items) float array of 0/1 responses. Missing cells may hold
        NaN / any value; they are ignored wherever ``observed`` is False.
    observed : (n_persons, n_items) bool array; True where a response is present.

    Returns
    -------
    dict with keys: ``a`` (discrimination), ``b`` (difficulty/intercept, so that
    logit = a*theta + b), ``theta`` (EAP ability), ``loglik_trace``, ``n_iter``,
    ``status``.
    """
    validated_nodes = _validate_quadrature_node_count(n_nodes)
    if isinstance(max_iter, (bool, np.bool_)) or not isinstance(
        max_iter, (int, np.integer)
    ):
        raise ValueError("max_iter must be an integer >= 1")
    validated_max_iter = int(max_iter)
    if validated_max_iter < 1:
        raise ValueError("max_iter must be an integer >= 1")

    y_array = np.asarray(y)
    observed_array = np.asarray(observed)
    if y_array.shape != observed_array.shape or y_array.ndim != 2:
        raise ValueError("y and observed must be 2D and identically shaped")

    n_persons, n_items = y_array.shape
    _validate_mmle_workspace(n_persons, n_items, validated_nodes)

    y = np.asarray(y_array, dtype=np.float64)
    observed = np.asarray(observed_array, dtype=bool)
    if not observed.any():
        raise ValueError("no observed responses")

    # Zero-fill missing so array math is finite; the observed mask nullifies them.
    y_filled = np.where(observed, y, 0.0)
    obs_f = observed.astype(np.float64)

    nodes, weights = gauss_hermite_nodes(validated_nodes)  # (Q,), (Q,)
    log_weights = np.log(weights)

    rng = np.random.default_rng(seed)
    # Init: a=1, b from observed item log-odds of endorsement.
    p_item = (y_filled * obs_f).sum(0) / np.clip(obs_f.sum(0), 1.0, None)
    p_item = np.clip(p_item, 0.02, 0.98)
    a = np.ones(n_items) + 0.01 * rng.standard_normal(n_items)
    b = np.log(p_item / (1.0 - p_item))

    nodes_sq = nodes * nodes
    loglik_trace: list[float] = []
    status = "max_iter_reached"

    for iteration in range(validated_max_iter):
        # ---- E-step: posterior over quadrature nodes per person ----
        # logit_{q,i} = a_i * node_q + b_i  ->  (Q, n_items)
        logit = nodes[:, None] * a[None, :] + b[None, :]
        log_p1 = -np.logaddexp(0.0, -logit)  # log sigmoid
        log_p0 = -np.logaddexp(0.0, logit)  # log(1 - sigmoid)
        # Per person, per node: sum over OBSERVED items of log P(y_pi | node_q)
        # log_lik[p, q] = sum_i obs_pi * (y_pi*log_p1_qi + (1-y_pi)*log_p0_qi)
        # Compute without forming n_persons * n_items * Q arrays via explicit broadcast products
        pos = (y_filled * obs_f) @ log_p1.T
        neg = ((1.0 - y_filled) * obs_f) @ log_p0.T
        log_joint = pos + neg + log_weights[None, :]  # + log prior weight
        # Normalize across nodes (log-sum-exp)
        max_lj = log_joint.max(axis=1, keepdims=True)
        stab = np.exp(log_joint - max_lj)
        denom = stab.sum(axis=1, keepdims=True)
        posterior = stab / denom  # (n_persons, Q)
        person_loglik = max_lj[:, 0] + np.log(denom[:, 0])
        total_loglik = float(person_loglik.sum())
        loglik_trace.append(total_loglik)

        # ---- M-step: update a, b by weighted logistic regression per item ----
        # Expected counts at each node: n_iq = sum_p obs_pi * posterior_pq  (Q per item)
        # r_iq = sum_p obs_pi * y_pi * posterior_pq
        # Compute without forming n_items * Q arrays via explicit broadcast products
        n_iq = obs_f.T @ posterior
        r_iq = (obs_f * y_filled).T @ posterior

        a_new = a.copy()
        b_new = b.copy()

        # Process active items together to remove the per-item Python control-flow loop.
        active = np.ones(n_items, dtype=bool)

        for _ in range(25):
            if not active.any():
                break

            ai = a_new[active]
            bi = b_new[active]
            n_iq_active = n_iq[active]
            r_iq_active = r_iq[active]

            eta_m = ai[:, None] * nodes[None, :] + bi[:, None]
            p_m = _sigmoid(eta_m)
            w = n_iq_active * p_m * (1.0 - p_m)
            resid = r_iq_active - n_iq_active * p_m

            # Optimization: Replace element-wise multiply and axis reduction with dense matrix multiplication
            # to avoid large intermediate array allocations
            g_a = resid @ nodes - ridge_a * ai
            g_b = resid.sum(axis=1) - ridge_b * bi

            h_aa = -(w @ nodes_sq) - ridge_a
            h_bb = -w.sum(axis=1) - ridge_b
            h_ab = -(w @ nodes)

            det = h_aa * h_bb - h_ab * h_ab

            valid = np.abs(det) >= 1e-12

            da = np.zeros_like(ai)
            db = np.zeros_like(bi)

            if valid.any():
                da[valid] = (h_bb[valid] * g_a[valid] - h_ab[valid] * g_b[valid]) / det[
                    valid
                ]
                db[valid] = (h_aa[valid] * g_b[valid] - h_ab[valid] * g_a[valid]) / det[
                    valid
                ]

            ai -= da
            bi -= db
            ai = np.clip(ai, 1e-3, 10.0)

            converged = (np.abs(da) + np.abs(db)) < 1e-8
            done = converged | ~valid

            a_new[active] = ai
            b_new[active] = bi

            active[active] = ~done

        a, b = a_new, b_new

        if iteration > 0 and abs(loglik_trace[-1] - loglik_trace[-2]) < tol:
            status = "converged"
            break

    # ---- EAP ability for each person ----
    # Optimization: Replace element-wise multiply and axis reduction with dense matrix multiplication
    # to avoid intermediate array allocation
    theta = posterior @ nodes

    return {
        "a": a,
        "b": b,
        "theta": theta,
        "loglik_trace": loglik_trace,
        "n_iter": len(loglik_trace),
        "status": status,
    }
